### PR TITLE
Custom vmap implementation

### PR DESCRIPTION
**Pull request by zou3519**
TODO: needs description of what is going on

### CODE REVIEW

File: simple_functorch.py [hunk 1]
```diff
@@ -740,6 +764,10 @@ def new_fn(d, *args):
         r, saved = self.inner.custom_vjp(batchify(fwd_fn), batchify(bwd_fn), *args)
         return r, saved
 
+    def custom_vmap(self, fn, batch_rule, *args):
+        result = batch_rule(self.inner, *args)
+        return result
```
Comment (zou3519) [inline simple_functorch.py:769]: Unlike custom_vjp, custom_vmap does not call custom_vmap on the inner dispatcher!

File: simple_functorch.py [hunk 2]
```diff
@@ -1277,3 +1305,38 @@ def run_gradvmap(d2: "Batched", d1: "Autograd"):
...
+
+# autograd should let custom_vmap pass through
+x = label(torch.tensor([[1., 2., 3.], [4., 5., 6.]]))
+d = Autograd(Batched(Torch(), length=2))
+result = d.custom_vmap(f, f_batch_rule, x)
+assert torch.allclose(result, 2 * x)
+loss88 = d.sum(result, name='loss88')
+
+grad_x, = d.grad(loss88, [x])
+assert torch.allclose(grad_x, torch.full_like(x, 2))
```
Comment (albanD) [inline simple_functorch.py:1342]: I don't think I agree with this?
I would expect that `Autograd(Batched(Torch(), length=2))` would give `2 * x` as the gradient while `Batched(Autograd(Torch()), length=2)` would give `2`. No?

Comment (albanD) [reply simple_functorch.py:1342]: Yes.
I would expect here to be able to see the different between batch of gradients and element-wise gradients.

Comment (albanD) [reply simple_functorch.py:1342]: Isn't that question beyond what we're discussing here?
The question is still there without considering any custom_vjp right?

Comment (zou3519) [reply simple_functorch.py:1342]: Oh I'm just trying to understand the difference between this and custom_vjp. I think the current semantics are analogous to what custom_vjp is doing but reasoning through it is confusing

Comment (zou3519) [reply simple_functorch.py:1342]: cc @ezyang @Chillee -- this is what we were discussing during the Composability hangout, if you folks have opinions

Comment (samdow) [reply simple_functorch.py:1342]: Disregarding custom_vjp for a moment, if we're going off of the argument from the meeting today that this should match the behavior of a normal batch rule, isn't this code right?

Let's say we're calling `d2.unsqueeze(x, dim)` with `d2 = Batched(Autograd(Torch()))`. First it hits the Batched dispatcher, so we get `d2.inner.unsqueeze(x, dim + 1)` so the Autograd dispatcher sees `unsqueeze(x, dim + 1)` (which is the batch rule) and does autograd on _that_ function. Using the same dispatcher stack, we similarly expect autograd runs on the "batch rule"

As a related a note, I always get confused that `Batched(Autograd(Torch()))` is the same as `grad(vmap())` so it might be worth to add the transform implementations if that's not too much work? I think this makes a lot more sense if we're able to say `grad(vmap())` gets the derivative of the custom batch rule rather than remembering to reason about the interpreter stack.

Comment (albanD) [reply simple_functorch.py:1342]: I agree that the current Dispatcher for the regular code does that.
But now if we call the `custom_vmap` with this unsqueeze function does it do that as well?

Comment (samdow) [reply simple_functorch.py:1342]: > But now if we call the custom_vmap with this unsqueeze function does it do that as well?

Yes?

More explanation:
> I would expect that Autograd(Batched(Torch(), length=2)) would give 2 * x as the gradient

Given that the original function is `d.mul(x, x)`, this would mean that autograd is running on the unbatched function, not the batch rule. If we agree that given this set of Dispatchers should end up with Autograd running on the batched rule, then 2 is the expected gradient because `d.add(x, x)` is the batched rule and the derivative of that is 2

File: simple_functorch.py [hunk 3]
```diff
@@ -510,6 +515,25 @@ def propagate(dL_doutputs: List[Tensor]):
         )
         return r, saved
 
+    def custom_vmap(self, fn, batch_rule, *args):
+        def call_with_current_dispatcher(fn):
+            def wrapped(d, *args):
+                saved = self.inner
+                try:
+                    self.inner = d
```
Comment (ezyang) [inline simple_functorch.py:523]: woooow so spicy

Comment (ezyang) [reply simple_functorch.py:523]: I was going to say that you can do this without mutating the dispatcher stack by simply creating a new Autograd dispatcher on the fly, whose inner is d, but then the tapes would not be shared.  This seems... dubious.

File: simple_functorch.py [hunk 4]
```diff
@@ -510,6 +515,25 @@ def propagate(dL_doutputs: List[Tensor]):
         )
         return r, saved
 
+    def custom_vmap(self, fn, batch_rule, *args):
```
Comment (ezyang) [inline simple_functorch.py:518]: This appears to diverge substantially from JAX's custom_vmap, at https://github.com/google/jax/blob/main/jax/_src/custom_batching.py

File: simple_functorch.py (hunk 3 above)
Comment (ezyang) [reply simple_functorch.py:523]: I guess if we represent the tape with an extra indirection this isn't too hard to do. Probably better and then makes this rule nicely symmetric for how `custom_vjp` is implemented in `Batched`.

File: simple_functorch.py (hunk 1 above)
Comment (ezyang) [reply simple_functorch.py:769]: how come. It doesn't seem to me like Batched(Batched()) wouldn't apply. Perhaps you are saying, it is impossible for a batching rule to recursively refer to itself?

File: simple_functorch.py (hunk 4 above)
Comment (zou3519) [reply simple_functorch.py:518]: @ezyang is your comment that the API is different, the implementation is different, or both?

Comment (ezyang) [reply simple_functorch.py:518]: implementation. But later I worked out that this is exactly analogous to how we did batching, so... idk, maybe it's still fine
